### PR TITLE
fix(coarse-tile): rebuild cross-group reads correctly

### LINKED ISSUE #4008: Cross group coarse tiling doesn't correctly rebuild dependency graph when reading results across groups
### What happened?

When using the reproducer. We expect to be able to reference tensors between tiling groups using HBM materializations of full intermediates. Instead, the result is numerically incorrect.

### What did you expect to happen?

I expected a spyre and cpu to match.

### How can we reproduce it?

```
# Temporary repro: two SEQUENTIAL coarse-tile hint groups chained through a
# value. On stage1 this gives ~94.6% mismatch. Used to check whether stage6
# fixes it.
import os
import sys

# The test driver lives in tests/inductor/; put it on the path so this script
# runs standalone from the repo root.
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "tests", "inductor"))

import torch

from test_coarse_tile_e2e import run_coarse_tile_test, tensor
from torch_spyre._inductor import spyre_hint


"""z = abs(x)+y under A/4 (group 1), then out = z*2 under a SEPARATE
A/4 scope (group 2). Two chained groups on [512,256]."""
inputs = [
    tensor("x", shape=(512, 256), dims=["A", "B"]),
    tensor("y", shape=(512, 256), dims=["A", "B"]),
]

def fn(x, y):
    with spyre_hint(num_tiles_per_dim={"A": 4}):
        with spyre_hint(expected_named_dims=["A", "B"]):
            z = torch.abs(x) + y
    with spyre_hint(num_tiles_per_dim={"A": 4}):
        with spyre_hint(expected_named_dims=["A", "B"]):
            out = z * 2
    return out

run_coarse_tile_test(fn, inputs, loopspec=None)
```

### Any environmental details we need to know?

No. 

### Anything else we need to know?

Executed on 772344

### Relevant log output

```Shell
Tensor-likes are not close!

Mismatched elements: 123989 / 131072 (94.6%)
Greatest absolute difference: 15.953125 at index (184, 59) (up to 0.1 allowed)
Greatest relative difference: inf at index (8, 111) (up to 0.1 allowed)
```

### PR DESCRIPTION
**Pull request by jturney**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [x] bug
- [ ] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Two chained coarse-tile hint groups produced 94.6% wrong results: the second group striding-read the first group's 128-row per-tile scratch instead of its full HBM materialization.
Out rows 0-127 held tile 3's data (the last tile the scratch was overwritten with) and the rest read past the buffer.
Two stacked defects produced the one symptom.

**1. Pass-ordering blind spot in the copy-out consumer patching** (`coarse_tile.py`, `_propagate_tiled_op`).
Pass 1 interposes a read-copy staging op that takes over a cross-group consumer's read of the scratch.
Pass 3 then patched consumers from the planning-time name list, which predates the staging op: the redirect landed on an op that no longer performed the read, the staging op kept draining the scratch, and the untouched full buffer was scheduled dead and freed.
The fix resolves consumers at transform time, as the union of current out-of-group readers (catches staging ops) and planned names that still read the buffer (preserves same-outer-group deferred reduction consumers such as softmax's `div`, which the group test alone would wrongly drop; `test_softmax_2d_512x256_dim1_B4` regressed under an out-of-group-only resolution and pinned this).
A debug log reports when the resolved set diverges from the planned one.

**2. `loop_info` aliasing in `insert_restickify`.**
The interposed restickify node shared the consumer's `CoarseTileInfo` object, so the per-read tile advance could not be transferred: both ops advanced, and a coarse-tiled consumer of a cross-group full buffer read the one-tile stage with a striding index and ran off its end.
The restickify node now gets its own copy: its read takes over the dep's advance (it strides the source), its output is per-iteration scratch that never advances, and the consumer's advance for that dep is cleared.
This restores the invariant every healthy staging chain already exhibits: readers of stages never advance.

#### Which issue(s) this PR is related to:

Fixes #4008

#### Special notes for your reviewer:

The issue's reproducer is added verbatim as `test_two_chained_groups_512x256_A4`.
Verified on device: the reproducer passes (was 123989/131072 mismatched), and `test_coarse_tile_e2e.py`, `test_coarse_tiling.py`, `test_restickify.py`, and `test_work_division_hint.py` pass 849/849 together.
The second defect was only exposed after fixing the first (tile 0 became correct while tiles 1-3 stayed wrong), so the two fixes are intentionally one commit: neither is sufficient alone for the cross-group case.

#### Does this PR introduce a user-facing change?

No. Chained coarse-tile hint groups now compute correct results.

#### Additional note:

The same two mechanisms are worth checking in the coarse-tiling stage4/stage6 restructures, which the issue was originally filed to probe.

### CODE REVIEW

**Comment by jturney** [comment]
The failing test appears to be a hardware flake.

**Review by dgrove-oss — APPROVED** [review]
Review

BLOCKERs

None.

SUGGESTIONs

None.

QUESTIONs

None.

Verification performed:
- Confirmed the transform-time consumer resolution in _propagate_tiled_op is a faithful re-derivation of the existing _find_outside_consumers_planned planning-time test (same outer-loop_group_id exclusion logic), applied to live op objects instead of a name snapshot — consistent with the file's established pattern rather than new ad hoc logic.
- Confirmed the "softmax div" same-outer-group exception is real, pre-existing, documented behavior (_consumers_reading_incomplete_reduction), validating why the fix unions planned names with actual current readers rather than using actual-readers alone.
- Confirmed CoarseTileInfo is a plain (non-frozen) @dataclass, so the old restick_buff.loop_info = op.loop_info genuinely aliased one mutable object between two ops — the described bug is real, not hypothetical.
- Verified the fix's copy.copy(consumer_li) + wholesale (not in-place) reassignment of tiled_dims_per_read/output_tiled_dims on both the copy and the original correctly decouples the two objects, confirmed with a standalone repro of the shallow-copy semantics.
- Confirmed the untouched fallback path (else: restick_buff.loop_info = consumer_li, when there's no per-level/per-read info to redistribute) is safe to leave aliased, since there's nothing in that case for the bug to corrupt.
- Confirmed test_two_chained_groups_512x256_A4 is a verbatim match of issue #4008's posted reproducer.
- Both modified test files (test_coarse_tile_e2e.py) are already registered via existing OOT configs with unlisted_test_mode: mandatory_success, so no new config entry is needed and the new test is automatically covered.
- DCO: substantive commit signed off; the three merge commits lack signoff, consistent with established precedent that merge commits are not a real gap.
- Mergeable (MERGEABLE, blocked only on pending required reviews).
- CI was still running at review time (3 checks pending, 0 failing) — worth a final glance before merge, but nothing seen so far suggests trouble.

The author's own "Additional note" flags that the same two mechanisms should be checked in the stage4/stage6 coarse-tiling restructures — a reasonable, already-acknowledged follow-up, not something this PR needs to also fix.

Verdict

APPROVE — a well-diagnosed, minimally-scoped fix for two genuinely stacked defects, with a verbatim regression test and correctness claims that check out under independent inspection of the surrounding code and data model.
